core/templates/dev/head/pages/exploration_editor/statistics_tab/LearnerActionRenderService.js

```diff
@@ -0,0 +1,171 @@
+        }
+        var answer = finalBlock[i].actionCustomizationArgs.answer.value;
+        var feedback = finalBlock[i].actionCustomizationArgs.feedback.value;
+        tableHTML +=
+          '<tr><td>' + answer + '</td><td>' + feedback + '</td></tr>';
+      }
+      tableHTML += '</table>';
+      return tableHTML;
+    };
+
+    var renderLearnerActionHTML = function(learnerAction) {
+      var actionType = learnerAction.actionType;
+      var custArgs = learnerAction.actionCustomizationArgs;
+      if (actionType === ACTION_TYPE_EXPLORATION_START) {
+        return renderExplorationStartActionHTML(custArgs.state_name.value);
+      } else if (actionType === ACTION_TYPE_EXPLORATION_QUIT) {
+        return renderExplorationQuitActionHTML(
+          custArgs.state_name.value, custArgs.time_spent_in_state_secs.value);
+      } else if (actionType === ACTION_TYPE_ANSWER_SUBMIT) {
+        interactionId = custArgs.interaction_id.value;
+        if (interactionId === 'Continue') {
+          return renderContinueButtonSubmitActionHTML(
+            custArgs.dest_state_name.value,
+            custArgs.time_spent_in_state_secs.value);
+        }
+        else if (custArgs.answer.value) {
+          return renderAnswerSubmitActionHTML(
+            custArgs.answer.value, custArgs.dest_state_name.value,
+            custArgs.time_spent_in_state_secs.value, custArgs.state_name.value
+          );
+        }
+      }
+    };
+
+    return {
+      renderFinalDisplayBlockForMISIssueHTML: function(block) {
+        var index = block.length - 1;
+        var stateName = block[index].actionCustomizationArgs.state_name.value;
+        var htmlString = '';
+        for (
+          var i = 0; block[i].actionType !== ACTION_TYPE_ANSWER_SUBMIT; i++) {
+          htmlString += renderLearnerActionHTML(block[i]);
+        }
+        htmlString +=
+          '<span class="learner-action">Submitted the following answers in ' +
+          'card "' + stateName + '"</span>';
+        htmlString += renderLearnerActionsTableForMultipleIncorrectIssue(block);
+        htmlString += renderLearnerActionHTML(block[index]);
+        return htmlString;
+      },
+      renderDisplayBlockHTML: function(block) {
+        var htmlString = '';
+        for (var i = 0; i < block.length; i++) {
+          htmlString += renderLearnerActionHTML(block[i]);
+        }
+        return htmlString;
+      },
+      getDisplayBlocks: function(learnerActions) {
+        var displayBlocks = [];
+        var lastIndex = learnerActions.length - 1;
+        var localBlock = [learnerActions[lastIndex]];
+        var stateName =
+          learnerActions[lastIndex].actionCustomizationArgs.state_name.value;
+        for (var i = lastIndex - 1; i >= 0; i--) {
+          var action = learnerActions[i];
+          if (action.actionCustomizationArgs.state_name.value !== stateName) {
+            stateName = action.actionCustomizationArgs.state_name.value;
+            if (localBlock.length < 4) {
```

Comment:
This loop block looks too complicated (especially since you've hard-coded the value `4` into it). Please split this up into well-named helper functions to make it clear what actually needs to happen.

Reply:
Added some comments and sub functions.

Reply:
Created helper object.